### ISSUE TITLE
Crash in constructor with Android 7.0

### ISSUE THREAD

**Issue by Noxalus** [fuyufjh/GraphicBuffer#5]
Hello everyone,

I'm using this project to instanciate a GraphicBuffer and use it to create an EGL image calling (with `glEGLImageTargetTexture2DOES `), and it works on devices with Android 6.0.

But it crashes into the GraphicBuffer constructor on devices with Android 7.0+, is it supposed to work for the last versions of Android?

**Comment by AndroidInternal** [comment]
The old GraphicBuffer constructor changed since API24.

**Comment by linjiansheng** [comment]
Any solution for API 24？

**Comment by sergeivolodin** [comment]
**TLDR: The working example for APIs 26-27 and 24-25 is here: https://github.com/chili-epfl/GraphicBuffer (see README). Code from the original repo by @fuyufjh works for API 23.**

@linjiansheng , I'm working on one right now. I have fixed the constructor (it now takes 7 arguments instead of 4, new ones can be dummy ones), I will add my code as a repo and add a link here this week with the step-by-step howto of how I managed to make it work.

Moreover, on `API 24` there is a major change: the private libraries are now impossible to use in your app, see https://developer.android.com/about/versions/nougat/android-7.0-changes , "NDK Apps Linking to Platform Libraries". This makes the process harder since it's then required to pack all of the required `.so` files along with your app (I'm working on that now)

There is another public alternative for `API 26` and higher: C++ wrapper `AHardwareBuffer` over `HardwareBuffer`: see https://developer.android.com/ndk/guides/stable_apis (look for AHardwareBuffer) and https://developer.android.com/reference/android/hardware/HardwareBuffer.

Despite the fact that the code compiles, it doesn't work on my `API 25` device, even if I pack the `libnativewindow.so` file from `API 26` with my app. The factory function gives a weird highly negative return code, and the pointer to object is a `nullptr`.

It seems that for API <= `23` the `GraphicBuffer` worked, and for API >= `26` the `HardwareBuffer` works (I guess). However, there seem to be no solution for `24` and `25`, so I'm trying to hack one together for a library which I develop.

Again, I will post my code when it starts working (or when I understand that I need help :)

**Comment by sergeivolodin** [comment]
**TLDR: The working example for APIs 26-27 and 24-25 is here: https://github.com/chili-epfl/GraphicBuffer (see README). Code from the original repo by @fuyufjh works for API 23.**

<s>TLDR: code fixup https://github.com/chili-epfl/GraphicBuffer/commit/f882c015c14e486e2cb01bf61c979b4db0a4b361 + make your app a system app on API 25 (Android 7)
OR use `AHardwareBuffer` on API 27 (Android 8)</s>

I've managed to (so far)
1. Obtain the pointer to new constructor [1]. In order to obtain the name of the function in `libui.so` I got one from my Android device and found the respective name using `objdump`. See the patch in my repo [2].
2. Build my app and run it with a crash message (cannot use `/vendor/lib/hw/gralloc.exynos5.so` which is private even in `25`).
3. Bundle all of the required `.so` files [4] with my app. This didn't work, apparently it's not OK to include own `libc`. It crashed on the library initialization stage.
4. Make the app a system app [3] (requires root), which **removes all restrictions on private libraries such as `libui.so`**
5. Run the code which gets the pointer to image.

Despite the fact that `lock()` sets my pointer to some sensible value, the call `lock()` **returns with code 2**, which is not what should happen, if I got it correctly.

At this moment the output image is garbage (input is a camera image) but the whole process is extremely fast:
<img width="240" alt="shader0" src="https://user-images.githubusercontent.com/1012270/43852667-1f7eadc8-9b3e-11e8-9d2d-97572badd733.png">

To sum up, to this point, I have managed to overcome the constructor issue and the libraries issue. My output has some garbage content. I'm still working on that problem and will update this comment if I get more impressive results.

Would be glad if someone could test my code to check if it's my fix for `GraphicBuffer` which is wrong or it's my usage of it which is wrong.

**UPDATE:** I've tried out an Android 8 API `27` device with AHardwareBuffer. It seems like I get the same garbage output, so probably it's my usage of `GraphicBuffer` which is incorrect and the fix in my repo is OK

![w20180809144658_shader0 png_white](https://user-images.githubusercontent.com/1012270/43901350-8d4764cc-9be7-11e8-8bfd-6381e8a3109a.png)


**UPDATE2:** After updating SDK, NDK, and Qt to latest versions, I have managed to obtain some results on API 27 device using AHardwareBuffer (this is actually what camera sees)
<img width="240" alt="20180809163546_shader0" src="https://user-images.githubusercontent.com/1012270/43906242-164c617c-9bf3-11e8-8c9f-69db6ccaec6e.png">
<img width="240" alt="shader1" src="https://user-images.githubusercontent.com/1012270/43906243-16762e44-9bf3-11e8-8c7b-b0665312637b.png">

[1] https://android.googlesource.com/platform/frameworks/native/+/jb-dev/libs/ui/GraphicBuffer.cpp
[2] https://github.com/chili-epfl/GraphicBuffer/commit/f882c015c14e486e2cb01bf61c979b4db0a4b361
[3] https://stackoverflow.com/questions/24641604/qt-application-as-system-app-on-android
[4] `libbacktrace.so libbase.so libbinder.so libc.so libc++.so libcutils.so libdl.so libgralloc.exynos5.so libhardware.so libion.so liblog.so liblzma.so libm.so libsync.so libui.so libunwind.so libutils.so`

**Comment by sergeivolodin** [comment]
**TLDR: The working example for APIs 26-27 and 24-25 is here: https://github.com/chili-epfl/GraphicBuffer (see README). Code from the original repo by @fuyufjh works for API 23.**

I have successfully obtained a rendered texture image on API 25 (Android 7, GraphicBuffer) and API 27 (Android 8, HardwareBuffer).

The code is different for these, but the idea is the same.

My old code (links above) contains mistakes and shouldn't work, use this version instead.

Full howto and the code is here: https://github.com/chili-epfl/GraphicBuffer
